pkg/storage/storage_tests/module_storage/storage_benchmark_test.go

```diff
@@ -0,0 +1,135 @@
+package modulestorage
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/gobuffalo/suite"
+	"github.com/gomods/athens/pkg/storage"
+	"github.com/gomods/athens/pkg/storage/fs"
+	"github.com/gomods/athens/pkg/storage/mem"
+	"github.com/gomods/athens/pkg/storage/minio"
+	"github.com/gomods/athens/pkg/storage/mongo"
+	"github.com/gomods/athens/pkg/storage/rdbms"
+	"github.com/stretchr/testify/require"
+)
+
+func BenchmarkStorageList(b *testing.B) {
+	module, version := "athens_module", "1.0.1"
+	zip, info, mod := []byte("zip_data"), []byte("info"), []byte("mod_info")
+
+	for _, store := range getStores(b) {
+
+		backend := store.Storage()
+
+		require.NoError(b, backend.Save(context.Background(), module, version, mod, bytes.NewReader(zip), info), "Save for storage %s failed", backend)
+
+		b.Run(fmt.Sprintf("listing module backend %s", store.StorageHumanReadableName()), func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				backend.List(context.Background(), module)
+			}
+		})
+
+		require.NoError(b, store.Cleanup())
+	}
+}
+
+func BenchmarkStorageSave(b *testing.B) {
+	module, version := "athens_module", "1.0.1"
+	zip, info, mod := []byte("zip_data"), []byte("info"), []byte("mod_info")
+
+	for _, store := range getStores(b) {
+
+		backend := store.Storage()
+
+		b.Run(fmt.Sprintf("save module backend %s", store.StorageHumanReadableName()), func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				err := backend.Save(context.Background(), fmt.Sprintf("%s-%d", module, i), version, mod, bytes.NewReader(zip), info)
+				require.NoError(b, err)
+			}
+		})
+
+		require.NoError(b, store.Cleanup())
+	}
+}
+
+func BenchmarkStorageDelete(b *testing.B) {
+	module, version := "athens_module", "1.0.1"
+	zip, info, mod := []byte("zip_data"), []byte("info"), []byte("mod_info")
+
+	for _, store := range getStores(b) {
+
+		backend := store.Storage()
+
+		for i := 0; i < b.N; i++ {
+			err := backend.Save(context.Background(), fmt.Sprintf("del-%s-%d", module, i), version, mod, bytes.NewReader(zip), info)
+			require.NoError(b, err, "save for storage %s module: %s failed", backend, i)
+		}
+
+		b.Run(fmt.Sprintf("delete module backend %s", store.StorageHumanReadableName()), func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				backend.Delete(context.Background(), fmt.Sprintf("del-%s-%d", module, i), version)
+			}
+		})
+
+		require.NoError(b, store.Cleanup())
+	}
+}
+
+func BenchmarkStorageExists(b *testing.B) {
+	module, version := "athens_module", "1.0.1"
+	zip, info, mod := []byte("zip_data"), []byte("info"), []byte("mod_info")
+
+	for _, store := range getStores(b) {
+
+		backend := store.Storage()
+
+		for i := 0; i < b.N; i++ {
+			err := backend.Save(context.Background(), fmt.Sprintf("%s-%d", module, i), version, mod, bytes.NewReader(zip), info)
```

Comment:
this `N` can be different than the `N` inside the `Run` function, furthermore `N` actually changes and its not constant depending on how long the operations take over each iteration. For this test, you don't need to create multiple modules. Just create 1 module and test `backend.Exists` N times over that one module.

Reply:
created two `b.Run` one for existing and other non-existing modules, would it be apt to pull it out as a separate benchmark?

for existing tests, we gotta save the modules, so i could pull out a `x := b.N` constant for saving x modules, and while doing `exists` could do  `n %  x`. would that work?

Reply:
@devdinu for existing to test you only need to save `1` module, not `N` module :) just check that this one module Exists N times.

Reply:
i was thinking for single module `same query` it would get `cached` and that wouldn't give us the right number.

Reply:
@devdinu yes load testing is separate, though very important for scalability. I'll create an issue for it. As for `cached`, that's Okay. We don't care how the backend implementation does things. We just wanna compare them against each other. For example, if Mongo caches things but GCP does not, then we wanna see that Mongo is actually faster on average.